segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadOperation.java

```diff
@@ -141,23 +143,56 @@ private void logEnd() {
+                        return CompletableFuture.runAsync(() -> {
+                            // Create parallel requests to read each chunk.
+                            chunkReadFutures.add(readChunk(chunkToReadFrom.getName(),
+                                    currentOffset.get() - startOffsetForCurrentChunk.get(),
+                                    bytesToRead,
+                                    currentBufferOffset.get()));
+                            log.trace("{} read - reading chunk - op={}, segment={}, chunk={} offset={} length={} bufferOffset={}",
+                                    chunkedSegmentStorage.getLogPrefix(),
+                                    System.identityHashCode(this), handle.getSegmentName(),
+                                    chunkToReadFrom.getName(),
+                                    currentOffset.get() - startOffsetForCurrentChunk.get(),
+                                    bytesToRead,
+                                    currentBufferOffset.get());
+                            bytesRemaining.addAndGet(-bytesToRead);
+                            currentOffset.addAndGet(bytesToRead);
+                            currentBufferOffset.addAndGet(bytesToRead);
+                            totalBytesRead.addAndGet(bytesToRead);
+                        }, chunkedSegmentStorage.getExecutor());
                     }
                     return CompletableFuture.completedFuture(null);
-                }, chunkedSegmentStorage.getExecutor());
+                }, chunkedSegmentStorage.getExecutor())
+                .thenCompose(v -> Futures.allOf(chunkReadFutures));
+    }
+
+    private CompletableFuture<Void> readChunk(String chunkName,
+                         long fromOffset,
+                         int bytesToRead,
+                         int bufferOffset) {
+        val chunkBytesRemaining = new AtomicInteger(bytesToRead);
+        val chunkFromOffset = new AtomicLong(fromOffset);
+        val chunkBufferOffset = new AtomicInteger(bufferOffset);
+        // Note that technically it is possible that read actually request reads less than requested bytes, requiring additional reads on the same chunk.
+        // Hence the for loop below.
+        return chunkedSegmentStorage.getChunkStorage().openRead(chunkName)
+                .thenComposeAsync(chunkHandle ->
+                    Futures.loop(
+                            () -> chunkBytesRemaining.get() > 0 && null != currentChunkName,
+                            () -> chunkedSegmentStorage.getChunkStorage().read(chunkHandle,
+                                    chunkFromOffset.get(),
+                                    chunkBytesRemaining.get(),
+                                    buffer,
+                                    chunkBufferOffset.get())
+                                    .thenAccept(n -> {
+                                        Preconditions.checkState(n != 0, "Zero bytes read");
```

Comment:
Add more information to this. If you see it in the logs, you will want to have as much info as you can have in order to debug. There's no chunk, no start/end offset, etc. in this message.

Reply:
fixed